MdeModulePkg/Library/UefiBootManagerLib/BmBoot.c

```diff
@@ -2436,6 +2436,241 @@ BdsLibGetBootableHandle (
+        EFI_IMAGE_MACHINE_TYPE_SUPPORTED (Hdr.Pe32->FileHeader.Machine) &&
+        Hdr.Pe32->OptionalHeader.Subsystem == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION) {
+        BootOptions = CreatePreInstalledBootOption(
+                        BootOptions,
+                        BootOptionCount,
+                        Handles[Index],
+                        &DtsBootOpt
+                        );
+      } else {
+        continue;
+      }
+    }
+
+    OptionsArraySize = sizeof (PreInstalledBootOpts) / sizeof(PRE_INSTALLED_BOOT_OPT);
+    for (OsIdx = 0; OsIdx < OptionsArraySize; OsIdx++) {
+
+      Hdr.Union = &HdrData;
+      Status = BdsLibGetImageHeader (
+                 Handles[Index],
+                 PreInstalledBootOpts[OsIdx].FileName,
+                 &DosHeader,
+                 Hdr
+                 );
+      if (!EFI_ERROR (Status) &&
+        EFI_IMAGE_MACHINE_TYPE_SUPPORTED (Hdr.Pe32->FileHeader.Machine) &&
+        Hdr.Pe32->OptionalHeader.Subsystem == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION) {
+        BootOptions = CreatePreInstalledBootOption(
+                        BootOptions,
+                        BootOptionCount,
+                        Handles[Index],
+                        &PreInstalledBootOpts[OsIdx]
+                        );
+      }
+
+    } // for PreInstalledBootOpts
+
+    OptionsArraySize = sizeof (PreInstalledBootOptsLnx) / sizeof(PRE_INSTALLED_BOOT_OPT);
+    for (OsIdx = 0; OsIdx < OptionsArraySize; OsIdx += 2) {
+
+      Hdr.Union = &HdrData;
+      Status = BdsLibGetImageHeader (
+                 Handles[Index],
+                 PreInstalledBootOptsLnx[OsIdx].FileName,
+                 &DosHeader,
+                 Hdr
+                 );
+      if (!EFI_ERROR (Status) &&
+        EFI_IMAGE_MACHINE_TYPE_SUPPORTED (Hdr.Pe32->FileHeader.Machine) &&
+        Hdr.Pe32->OptionalHeader.Subsystem == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION) {
+        BootOptions = CreatePreInstalledBootOption(
+                        BootOptions,
+                        BootOptionCount,
+                        Handles[Index],
+                        &PreInstalledBootOptsLnx[OsIdx]
+                        );
+      } else {
+        /* Shimx64.efi not found or any other error, try grubx64.efi */
+        Hdr.Union = &HdrData;
+        Status = BdsLibGetImageHeader (
+                 Handles[Index],
+                 PreInstalledBootOptsLnx[OsIdx + 1].FileName,
+                 &DosHeader,
+                 Hdr
+                 );
+        if (!EFI_ERROR (Status) &&
+          EFI_IMAGE_MACHINE_TYPE_SUPPORTED (Hdr.Pe32->FileHeader.Machine) &&
+          Hdr.Pe32->OptionalHeader.Subsystem == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION) {
+          BootOptions = CreatePreInstalledBootOption(
+                          BootOptions,
+                          BootOptionCount,
+                          Handles[Index],
+                          &PreInstalledBootOptsLnx[OsIdx + 1]
+                          );
+        }

```

Comment:
This block of code repeats twice in this loop and two more times above.  Consider moving it to a function or moving file check inside of `CreatePreInstalledBootOption` which won't even need a new parameter, but might need signature change to return boolean and take `BootOptions` by reference as this loop needs branching.

Reply:
Created a new function that simplifies the flow: https://github.com/Dasharo/edk2/pull/70/commits/d9d411152e4b6244f44d4f5af663b63145b6e4f6#diff-d2820fe49b39a17c1aa46361e3d22ee285be3bd600dae98c0dc36d3bf030d04fR2614